### DISCUSSION TITLE
What is required to get Heirarchical transforms to work?

### DISCUSSION THREAD

**Discussion by sdfgeoff** [bevyengine/bevy#1569] [Q&A] [answered]
Originally I was going to post this as an issue, but it's probably more likely to be an issue on my side. But I've still followed the issue template.

## Bevy version

Recent git:
source = "git+https://github.com/bevyengine/bevy?branch=main#d9fb61d474518eba06d0c3dce5cb81db9f7ed66a"

## Operating system & version

Linux

## What you did

Set up a bunch of heirarchies of objects and meshes, loaded from scenes

## What you expected to happen

When you transform a parent, the children should also be transformed such that they act like a single "unit"

## What actually happened

The results are inconsistent. Sometimes no objects are updated, sometimes some objects are updated

## Additional information

Consider this scene file:
```
[
    (entity: 0, components:[
        {"type":"blend_loader::blend_mesh::BlendMeshLoader","struct":{"path":"Meshes/26ac592701faecdf9c9ba136dbe087b2.mesh"}},
        {"type":"blend_loader::blend_label::BlendLabel","struct":{"name":"Ship"}},
        {"type":"bevy_transform::components::transform::Transform","struct":{"translation":{"type":"glam::vec3::Vec3","value":(0.0,0.0,0.0)},"rotation":{"type":"glam::quat::Quat","value":(0.0,0.0,0.0,1.0)},"scale":{"type":"glam::vec3::Vec3","value":(1.0,1.0,1.0)}}},
        {"type":"bevy_transform::components::global_transform::GlobalTransform","struct":{"translation":{"type":"glam::vec3::Vec3","value":(0.0,0.0,0.0)},"rotation":{"type":"glam::quat::Quat","value":(0.0,0.0,0.0,1.0)},"scale":{"type":"glam::vec3::Vec3","value":(1.0,1.0,1.0)}}},
        {"type":"bevy_transform::components::children::Children","tuple_struct":[{"type":"smallvec::SmallVec<[bevy_ecs::core::entities::Entity; 1]>","list":[{"type":"bevy_ecs::core::entities::Entity","value":1}]}]}
    ]),
    (entity: 1, components:[
        {"type":"blend_loader::blend_mesh::BlendMeshLoader","struct":{"path":"Meshes/26ac592701faecdf9c9ba136dbe087b2.mesh"}},
        {"type":"blend_loader::blend_label::BlendLabel","struct":{"name":"Cube.002"}},
        {"type":"bevy_transform::components::transform::Transform","struct":{"translation":{"type":"glam::vec3::Vec3","value":(0.0,-5.0,0.0)},"rotation":{"type":"glam::quat::Quat","value":(-0.7071068286895752,0.0,0.0,0.7071068286895752)},"scale":{"type":"glam::vec3::Vec3","value":(1.0,1.0,1.0)}}},
        {"type":"bevy_transform::components::global_transform::GlobalTransform","struct":{"translation":{"type":"glam::vec3::Vec3","value":(0.0,-5.0,0.0)},"rotation":{"type":"glam::quat::Quat","value":(-0.7071068286895752,0.0,0.0,0.7071068286895752)},"scale":{"type":"glam::vec3::Vec3","value":(1.0,1.0,1.0)}}},
        {"type":"bevy_transform::components::children::Children","tuple_struct":[{"type":"smallvec::SmallVec<[bevy_ecs::core::entities::Entity; 1]>","list":[{"type":"bevy_ecs::core::entities::Entity","value":2}]}]},
        {"type":"bevy_transform::components::parent::Parent","tuple_struct":[{"type":"bevy_ecs::core::entities::Entity","value":0}]}
    ]),
    (entity: 2, components:[
        {"type":"blend_loader::blend_mesh::BlendMeshLoader","struct":{"path":"Meshes/26ac592701faecdf9c9ba136dbe087b2.mesh"}},
        {"type":"blend_loader::blend_label::BlendLabel","struct":{"name":"Cube.003"}},
        {"type":"bevy_transform::components::transform::Transform","struct":{"translation":{"type":"glam::vec3::Vec3","value":(0.0,-5.0,0.0)},"rotation":{"type":"glam::quat::Quat","value":(0.0,0.0,0.0,1.0)},"scale":{"type":"glam::vec3::Vec3","value":(1.0,1.0,1.0)}}},
        {"type":"bevy_transform::components::global_transform::GlobalTransform","struct":{"translation":{"type":"glam::vec3::Vec3","value":(0.0,-5.0,5.0)},"rotation":{"type":"glam::quat::Quat","value":(-0.7071068286895752,0.0,0.0,0.7071068286895752)},"scale":{"type":"glam::vec3::Vec3","value":(1.0,1.0,1.0)}}},
        {"type":"bevy_transform::components::parent::Parent","tuple_struct":[{"type":"bevy_ecs::core::entities::Entity","value":1}]}
    ])
]
```

There are three objects with parent/child relations set up between them ("Cube.003" is child of "Cube.002" is child of "Ship"). I have a system that rotates any entity called "Ship" - which for the above scene is the root of the hierarchy:
```rust
fn test_motion_system(mut query: Query<(Mut<Transform>, &BlendLabel)>) {
    for (mut transform, label) in query.iter_mut() {
        if label.name == "Ship" {
            println!("label: {}", label.name);
            transform.rotate(Quat::from_axis_angle(Vec3::new(0.0, 0.5, 0.866), 0.05))
        }
    }
}
```

My expectation is that, given that scene file, the objects would all rotate. But:

https://user-images.githubusercontent.com/13490050/110186861-64135d80-7e7b-11eb-9bde-48dd045b341d.mp4

However, that isn't to say it doesn't work. Another scene file - a much larger one, produces this result when rotated:

https://user-images.githubusercontent.com/13490050/110186949-a76dcc00-7e7b-11eb-9d60-745521e9b2d1.mp4

As you can see, many of the child entities are being rotated, but some are not. Also, of the ones that are being rotated, some of them are in the wrong location. The asset should look like:
![image](https://user-images.githubusercontent.com/13490050/110187144-2c58e580-7e7c-11eb-8ede-fbcd102e5986.png)

My guess at the moment is that something is wrong with my scene files and it's messing with the transforms. As far as I can guess, all I need is Transform, Parent and Child components but perhaps there is another component I need?

**Comment by cart** [comment]
If you switch to Bevy 0.4 (instead of `main`), do you still have issues? This might be related to our new parallel system executor we added on `main`, which works as expected so far, but requires some "order of operations" fixes in internal bevy plugins. We're in the process of fixing those now.

**Reply by sdfgeoff** [reply]
I didn't manage to get scene loading with children to work in `0.4.0`. as per https://github.com/bevyengine/bevy/discussions/1549 
So I can't do an exact test easily. 
I'll have a shot at making some reproduction examples of this (but it may only be next weekend when I have time)

**Reply by Davier** [reply]
In my tests on `main`, when loading a hierarchy with both `Parent` and `Children` components set, I get this panic:
```
thread 'Compute Task Pool (0)' panicked at 'children already added', /home/user/.cargo/git/checkouts/bevy-f7ffde730c324c74/f61e44d/crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:53:13
```
If the scene only sets `Children` components, `Parent` components are not added automatically.
Finally, if the scene only sets `Parent` components, the `Children` components are added correctly.

**Reply by sdfgeoff** [reply]
> In my tests on `main`, when loading a hierarchy with both `Parent` and `Children` components set, I get this panic:
> 
> ```
> thread 'Compute Task Pool (0)' panicked at 'children already added', /home/user/.cargo/git/checkouts/bevy-f7ffde730c324c74/f61e44d/crates/bevy_transform/src/hierarchy/hierarchy_maintenance_system.rs:53:13
> ```
> 
> If the scene only sets `Children` components, `Parent` components are not added automatically.
> Finally, if the scene only sets `Parent` components, the `Children` components are added correctly.

I updated main again, and had the same result. After removing the Children components from the scene file, it loaded and worked fine. I guess I just got a bad commit :confused: 

Now I just need to figure out how to spawn a dynamic scene as a child and I'll have everything I need to actually start making my game.

**Reply by Davier** [reply]
I discovered meanwhile that the issue with having `Children` can also be avoided by adding both `Parent` and `PreviousParent`.

>Now I just need to figure out how to spawn a dynamic scene as a child and I'll have everything I need to actually start making my game.

One way is to convert it to a `Scene` and use `SceneSpawner::spawn_as_child()` as discussed in #1446.

**Reply by sdfgeoff** [reply]
> One way is to convert it to a `Scene` and use `SceneSpawner::spawn_as_child()` as discussed in #1446.

Unfortunately current master doesn't seem to have a `DynamicScene::get_scene()` method anymore....

**Reply by Davier** [reply]
I missed that. I think you can call `DynamicScene::write_to_world()` on a new `Scene`'s world to get the same effect.

**Comment by sdfgeoff** [comment, marked as answer]
Updating to a newer `main` fixed it.
